error-codes/README.md

```diff
@@ -0,0 +1,105 @@
+# Error Codes
+
+| Lifecycle Stage | Maturity      | Status | Latest Revision |
+| --------------- | ------------- | ------ | --------------- |
+| 1A              | Working Draft | Active | r0, 2023-01-23  |
+
+Authors: [@sukunrt]
+
+Interest Group: [@marcopolo], [@achingbrain]
+
+[@MarcoPolo]: https://github.com/MarcoPolo
+[@achingbrain]: https://github.com/achingbrain
+
+## Introduction
+
+When closing a connection or resetting a stream, it's useful to provide the peer
+with a code that explains the reason for the closure. This enables the peer to
+better respond to the abrupt closures. For instance, it can implement a backoff
+strategy to retry _only_ when it receives a `RATE_LIMITED` error code. An error
+code doesn't always indicate an error condition. For example, a node can terminate an idle connection, or close a connection because a connection to the same peer over a better transport is available. In both these cases, it can signal an appropriate error code to the other end. 
+
+## Semantics
+Error codes are unsigned 32-bit integers. The range 0 to 10000 is reserved for
+libp2p errors. Application specific errors can be defined by protocols from
+integers outside of this range. Error Codes can be signaled on Closing a connection or on resetting a Stream.   
+
+From an application perspective, error codes provide a best effort guarantee. On resetting a libp2p stream or closing a connection with an error code, the error code may or may not be delivered to the application on the remote end. The specifics depend on the transport used. For example, WebTransport doesn't support error codes at all, while WebRTC doesn't support Connection Close error codes, but supports Stream Reset error codes. 
+
+### Connection Close and Stream Reset Error Codes
+Error codes are defined separately for Connection Close and Stream Reset. The namespace doesn't overlap as it is clear from the context whether the stream was reset by the other end, or it was reset as a result of a connection close. 
+Implementations MUST provide the Connection Close error code on streams that are reset as a result of remote closing the connection. 
+
+Libp2p streams are reset unilaterally, calling `Reset` on a stream resets both the read and write end of a stream. For transports, like QUIC, which support cancelling the read and write ends of the stream separately, implementations MAY provide the ability to signal error codes separately on resetting either end. 
+
+## Error Codes Registry
+Libp2p connections are shared by multiple applications. The same connection used in the dht may be used for gossip sub, or for any other application. Any of these applications can close the underlying connection on an error, resetting streams used by the other applications. To correctly distinguish which application closed the connection, Connection Close error codes are allocated to applications from a central registry. 
```

Comment:
We should outline a process to allocate these error codes. e.g. something as simple as:

- A user requests a certain number of error codes.
- We ensure it's a reasonable request.
  - Verify the number is reasonable (e.g. not 16k errors)
  - Verify there is no conflict.
- Allocate that in a high number space.

The process exists so that we can make it easier for future reviewers.